### DISCUSSION TITLE
Architecture Thought

### DISCUSSION THREAD

**Discussion by lawweiliang** [AppFlowy-IO/AppFlowy#995] [General]
![image](https://user-images.githubusercontent.com/34936929/188714225-cb2e2946-9735-468f-b604-c4fd605dfa8a.png)

Hihi Friends,

This is the high-level architecture that I am thinking of for a new project. Just want to share it with you yeah. 

The whole architecture will use
Tauri + React/Vue (Frontend)
Rocket(API)

Reasons I am using React/Vue as UI framework
* Easy to find developer working on this
* Easy for clients to port from their current UI component to a new framework without rewriting everything again from scratch (save cost + save time)

Reasons I am using Tauri
* Easy for me to port the UI into different Operating Systems.

Reasons I am using Rocket 
* Pass those architecture's hurdles to Rocket( such as Routing, ORM and etc)
* Future if Rocket updates, I can update accordingly.

My expectation
* High-performance upgrade (at least 10x faster) (If it isn't, there is no point to move it to Rust)

---

From an architectural design perspective, Liang usually focuses on 3 key components. Update + Scale + Standard.

Update
Question: The system needs to update from time to time. How do I update the code? 
I pass this responsibility to the framework. If the framework upgrades, I upgrade as well. 

Scale
Question: What if traffic increase dramatically? Can I scale the system?
That is the reason why I usually split the system into two parts. Frontend and backend. (lots of work, yeah I got you, look at my hair over the profile picture, then you will figure out why.  😃 )
Using Nginx + Reddis + Database Partition to reduce loading pressure + etc.  

Standard
Question: How to maintain the standard of a system that every programmer will follow? 
The framework comes in handy, everyone follows framework standards. Buddy, trust me. this is going to save you a ton of time. 

---

Notes, This is just my thought yeah.  It might have some bias or angle that I do not think of. Feel free to point it out. 😃

**Comment by annieappflowy** [comment]
Hi @lawweiliang , thanks for sharing your new project with us. Can you please provide us with more context on the business side? This will help us better understand the needs behind your technical design.

**Reply by lawweiliang** [reply]
Currently, I try to implement it in multi-level marketing projects first. Would like to see how Rust improves the bonus calculation part.

Still, try and errors. Might be failed. If it failed, I will turn it back to famous frameworks like Next or Laravel.

There is only two business use case on my site. One is to create a high-frequency trading system, the other one is the bonus calculation of the Multi-level marketing system.
